ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -625,8 +625,8 @@ public Optional<TransactionReceiptWithMetadata> transactionReceiptByTransactionH
     // on a missing optional is appropriate.
     final TransactionLocation location = maybeLocation.get();
     final Block block = blockchain.getBlockByHash(location.getBlockHash()).orElseThrow();
-    final Transaction transaction =
-        block.getBody().getTransactions().get(location.getTransactionIndex());
+    final List<Transaction> transactions = block.getBody().getTransactions();
+    final Transaction transaction = transactions.get(location.getTransactionIndex());
```

Comment:
As a "fix it now" patch this seems like a reasonable solution.  However there will be performance implications with large blocks.  I think a longer term solution would be to add the log index into the `TransactionLocation` and save that when calculating the location.  The data can be written so the value may be missing, and when it is we use this logic.  Creating a seamless transition.  We also could add a DB migration to re-build the transaction locations in the background.

Reply:
How to initiate re-building transaction locations?

Reply:
@fab-10

Reply:
@Wetitpig `migrateVariables` was much easier to implement, since it only had to move few keys around, and did it synchronously, while for this re-build an asynchronous tasks is needed, with all the consequences of making it not hurting the overall Besu performance. 

My proposal for this is to:
1.  first support saving the location for newly imported blocks
2.  then for already imported blocks, that miss this data, calculate it and save on demand

Have not yet looking into the code, to understand how much it, but would like to hear what you think.

Reply:
Given that most transaction receipts may not be queried for multiple times I am not sure if it is worth to save the `logIndexOffset` for each transaction.
Probably I am going to try removing `parallel()` and see how much the performance impact really is.

Reply:
if calculating that data at runtime is good enough for the moment, we can postpone the work to persist that data until more performance is needed.